### ISSUE TITLE
CPU random number generator is slow

### ISSUE THREAD

**Issue by ailzhang** [pytorch/pytorch#22414]
## To Reproduce
```
import torch

from time import perf_counter

def run():
    ys = 0
    with torch.no_grad():
        for _ in range(100):
            ys = ys + torch.rand(256, 20).mm(torch.rand(20, 20)).mean()
            # ys = ys + np.random.rand(256, 20).dot(np.random.rand(20, 20)).mean()

def main():
    for _ in range(100):
        run()
    t0 = 0
    t0 -= perf_counter()
    for _ in range(1000):
        run()
    t0 += perf_counter()
    print(f"time = {t0 / 1000:.6f}")

if __name__ == '__main__':
    main()
```

From @colesbury 's investigation, random number generator is taking ~40% of total time. 
![cpu_slow](https://user-images.githubusercontent.com/5248122/60468423-65a07300-9c0e-11e9-9bf9-ff75b1ff167e.jpg)

cc: @ezyang @roosephu

**Comment by ezyang** [comment]
cc @syed-ahmed since our plan on record is to replace CPU RNG with Philox

**Comment by syed-ahmed** [comment]
A small update just to keep people in the loop:

@colesbury added an AVX PhiloxSIMD, xoshiro256**, and pcg64 benchmark: https://github.com/colesbury/benchmark-rngs. The Pros/Cons of changing cpu random engine to philox are:

Pros:
- @colesbury 's PhiloxSIMD is about 2x faster than mt19937
- Jumping ahead in random sequence is essentially free. For PCG64, jumping ahead is O(log n)
- Since jumping ahead is fast, we can increment state before launching a kernel, lock scope can be reduced and multithreading performance will be improved (https://github.com/pytorch/pytorch/issues/21731)
- CUDA kernels uses Philox, which means, we should be getting consistent distributions between both of these backends

Cons:
- CPU distributions will run slower for CPUs that don't support AVX2
- xoshiro256**, and pcg64 are faster than Philox, without even any AVX2 code

Based on the above, we have some questions to answer:
- Are we fine with keeping different rng engines for different backends? If the answer is yes, then pcg64 or xoshiro256** might be the better engine for cpu side.
- Should we work towards reducing the lock scope, which is currently for the entire kernel duration? If the answer is yes, Philox seems to be the better choice there.
- Can we optimize the jump ahead function for PCG64, such that incrementing its state is free before launching a kernel? https://github.com/syed-ahmed/benchmark-rngs/blob/69daf56d4681fd002770c477af1785d6e0fe37ef/PCG.h#L28-L43

**Comment by soumith** [comment]
all roads are pointing to switching the CPU to Philox

**Comment by colesbury** [comment]
> all roads are pointing to switching the CPU to Philox

That's not immediately clear. There is one significant upside, which is that jumping ahead is cheaper. This may outweigh other considerations (it's generally slower), but it's not immediately obvious. Jumping ahead is ~1ns on Philox and ~50-100ns on PCG64. (and xoroshiro256** is similar). For comparison, operator overhead is ~1000-2000 ns currently.

It's not clear what we gain by having the same engine for CPU and GPU. The rest of the distributions code and other CPU/GPU ops don't necessarily compute the same exact result, so it's not like you'll get the same results when switching between CPU and GPU code.

If we use the same engine for CPU and GPU but continue  having two instances of the engine (one for CPU and one for GPU) we run the risk of correlated streams, which would be really bad. Using one shared instance fixes that, but introduces its own complexities.

**Comment by ezyang** [comment]
I agree with Sam here.  If we want to make CPU and GPU line exactly,
that's not an unreasonable goal, but let's be serious about what extra
work we'd have to do here.  (There are partial wins here too: you could
just arrange for CPU and CUDA to consume the same number of randoms
for equivalent operations--this should be easy).

Excerpts from Sam Gross's message of 2019-07-09 10:42:02 -0700:
> > all roads are pointing to switching the CPU to Philox
> 
> That's not immediately clear. There is one significant upside, which is that jumping ahead is cheaper. This may outweigh other considerations (it's generally slower), but it's not immediately obvious. Jumping ahead is ~1ns on Philox and ~50-100ns on PCG64. (and xoroshiro256** is similar). For comparison, operator overhead is ~1000-2000 ns currently.
> 
> It's not clear what we gain by having the same engine for CPU and GPU. The rest of the distributions code and other CPU/GPU ops don't necessarily compute the same exact result, so it's not like you'll get the same results when switching between CPU and GPU code.
> 
> If we use the same engine for CPU and GPU but continue  having two instances of the engine (one for CPU and one for GPU) we run the risk of correlated streams, which would be really bad. Using one shared instance fixes that, but introduces its own complexities.
>

**Comment by syed-ahmed** [comment]
>The rest of the distributions code and other CPU/GPU ops don't necessarily compute the same exact result, so it's not like you'll get the same results when switching between CPU and GPU code.

We can align CPU and GPU closer by turning these distributions into C10_HOST_DEVICE and using them in the CUDA distribution kernels: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/DistributionsHelper.h. The only exception is curand_poisson (for the time being, we can keep using it and later if need be we can re-write a GPU friendly poisson kernel).

>If we use the same engine for CPU and GPU but continue having two instances of the engine (one for CPU and one for GPU) we run the risk of correlated streams, which would be really bad. Using one shared instance fixes that, but introduces its own complexities.

Interesting point! Shared Philox state is doable since it's just a pair of two uint64_t-s and we can move it to the Generator base class. The complex part as I understand is:
- how to increment the shared philox offset in the CPU side? CUDA increments the offset to the number of randoms that will be needed by each thread. How do we translate this to CPU? In other words, for the CUDA side, the `subsequence` parameter for a Philox constructor is the thread index - what would this be for the CPU? 0 when it's single threaded? what happens during multi threaded?
- @colesbury 's PhiloxSIMD returns the randoms in a different order. So that adds a little more thought to the previous questions. May be the CUDA Philox can also be made to return randoms in this order and there won't be a performance impact?

**Comment by colesbury** [comment]
> We can align CPU and GPU closer by turning these distributions into C10_HOST_DEVICE and using them in the CUDA distribution kernels

This can give exact results for simple distributions like integers and uniform floats on [0, 1), but starts to fall apart when you get to more complicated distributions or use any subsequent functions (gemm, softmax, etc.). For example, randn() uses sin which may differ slightly between implementations.

(See https://docs.nvidia.com/cuda/floating-point/index.html#considerations-for-heterogeneous-world for a summary)

I'm not sure what we gain by producing the same random numbers on CPU and GPU if subsequent computations won't produce the same results. (It's unlikely to give us bitwise reproducibility between CPU and GPU in any real application). Behaviors that are almost, but not exactly the same, can be worse than behaviors that are obviously different because they encourage incorrect assumptions.
